### PR TITLE
Spec requiring tax category when creating products

### LINKED ISSUE #11212: [DFC] Can't create products for instances requiring tax categories on products
## Description

Follow-up from #11145.

We need to be able to create products for instances which require `tax_category` on products; this currently works if this requirement is disabled.

## Expected Behavior

Creating a product via DFC should be possible for instances which require a tax category over products.

## Actual Behaviour

Creating a product via DFC currently fails if the instance requires a tax category.

## Steps to Reproduce

As a super admin
1. Visit /admin/tax_settings/edit
2. Enable `products require tax category` option and click Update.
3. Follow the steps on "What to test" in PR #11145
4. See that you get an error 500:
https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/64b012c9492eca0007c8ff84?event_id=64b012c900bed74715b20000&i=sk&m=nw

## Your Environment

* Version used:
* Browser name and version:
* Operating System and version (desktop or mobile):

## Possible Fix

I think we just need to add a tax category when creating the product:
https://github.com/openfoodfoundation/openfoodnetwork/blob/774bb7a6072d488056ed3886a70f419f2c3ad56e/engines/dfc_provider/app/services/supplied_product_builder.rb#L19-L29

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

- Closes #11212 

It doesn't really fix the error described in the issue but it shows that the error was caused by a misconfiguration of the instance.

Instead I discovered some unnecessary spec code and cleaned up a little.

#### What should we test?

- Create a product.
- Create a variant.
- The above with and without requiring a tax category.

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: spec/models/spree/variant_spec.rb [hunk 1]
```diff
@@ -23,20 +23,35 @@
...
-        end
+      it "loads the default tax category" do
+        default = create(:tax_category, is_default: true)
+
+        expect(variant.tax_category).to eq default
+        expect {
+          variant.tax_category = nil
+        }.to_not change {
+          variant.tax_category
+        }
```
Comment (rioug) [inline spec/models/spree/variant_spec.rb:42]: I am not following what we are testing here ? Are we just checking that a tax category can't be set to null if a default one exists ?

**Review by rioug — APPROVED** [review]
Looks good !

File: spec/models/spree/variant_spec.rb (hunk 1 above)
Comment (mkllnk) [reply spec/models/spree/variant_spec.rb:42]: > Are we just checking that a tax category can't be set to null if a default one exists ?

Yes, exactly. Well, I don't actually know what would be saved to the database and it doesn't really matter but the getter method always returns the default in this case.

**Review by dacook — APPROVED** [review]
Great, thanks for cleaning up! 🧹

**Comment by drummer83** [comment]
Hmm, sorry, I think I am missing something here @mkllnk. On all three staging servers I get an error 404:

```
404 Undocumented | Error: response status is 404Response headers cache-control: no-cache  content-encoding: br  content-security-policy: default-src 'self' https:; font-src 'self' https: data: fonts.gstatic.com; img-src 'self' https: data: *.s3.amazonaws.com; object-src 'none'; frame-ancestors 'none'; script-src 'self' https: 'unsafe-inline' 'unsafe-eval' *.stripe.com openfoodnetwork.innocraft.cloud maps.googleapis.com maps.gstatic.com d2wy8f7a9ursnm.cloudfront.net; style-src 'self' https: 'unsafe-inline' fonts.googleapis.com cdnjs.cloudflare.com; connect-src 'self' https: https://staging.coopcircuits.fr wss://staging.coopcircuits.fr  content-type: application/json  date: Thu,18 Jan 2024 13:22:13 GMT  referrer-policy: strict-origin-when-cross-origin  server: nginx  strict-transport-security: max-age=63072000; includeSubDomains  x-content-type-options: nosniff,nosniff  x-download-options: noopen  x-firefox-spdy: h2  x-permitted-cross-domain-policies: none  x-request-id: ecd5649a-b9f2-4f0a-b43e-1fd08a29ffd8  x-runtime: 0.204090  x-xss-protection: 1; mode=block,1; mode=block
-- | --
```
image

Could you please provide some hints?
What do I have to change in the given example to make a correct request?
Anything else I need to enable on the server or in the super admin area?

Thanks!

**Comment by mkllnk** [comment]
I'm not exactly sure what you did, but here are some hints:

* One of your screenshots contains an API key. Probably best to replace that one.
* The shown example contains a spree_product_id which is used to create a variant for an existing product. But the id looks like the example id. If it can't find that product then it would reply with 404. You can just take that out to create a new product.
* If you enabled requiring a tax category on the staging server, you need to have a default tax category.

**Comment by mkllnk** [comment]
My test notes above were aiming at the normal admin UX though. I wanted to be sure that I didn't break anything. The API itself is well covered with specs. I'm happy for you to test it as well but I thought it may be too much effort.

**Comment by drummer83** [comment]
> * The shown example contains a spree_product_id which is used to create a variant for an existing product. But the id looks like the example id. If it can't find that product then it would reply with 404.

Ah, thanks! That's probably it! I checked that the id wasn't already taken and thought that the new variant would get this id assigned. Thanks for clarification! I will try this within the next days - weekend is coming. :wink:

**Comment by drummer83** [comment]
Hi @mkllnk,

I have tested this now.

## After staging

| Tax cat. required | Default tax cat. available | Result |
| - | - | - |
| No | No | :heavy_check_mark: |
| Yes | No | Error 500 :heavy_check_mark: |
| Yes | Yes | :heavy_check_mark: |

Both have been tested for variants as well as for products.

According to your [comment here](https://github.com/openfoodfoundation/openfoodnetwork/issues/11212#issuecomment-1888504024) it is ok to see the error when a tax category is required but no default tax category is set. :+1: 

Great! This one is ready to go! :tada: 
Merging! :rocket:

**Pull request merged**